query.go

```diff
@@ -258,17 +375,36 @@ func (q *query) queryPeer(p peer.ID) *queryResult {
 
 		// add their addresses to the dialer's peerstore
 		q.dht.peerstore.AddAddrs(next.ID, next.Addrs, pstore.TempAddrTTL)
-		closer := q.localPeers.Add(next.ID)
-		foundCloserPeer = foundCloserPeer || closer
+		saw = append(saw, next.ID)
 	}
 
-	// did the successful query RPC fulfill the query stop condition ?
-	if q.stopFn(q.localPeers) {
-		q.cancel()
-	}
+	ch <- &queryUpdate{seen: saw, queried: []peer.ID{p}}
+}
 
-	return &queryResult{
-		foundCloserPeer: foundCloserPeer,
+func (q *query) updateState(up *queryUpdate) {
+	for _, p := range up.seen {
+		if p == q.dht.self { // don't add self.
+			continue
+		}
+		q.queryPeers.TryAdd(p)
```

Comment:
@petar you had `q.queryPeers.SetState(p, qpeerset.PeerHeard)` under here before. However, this allowed for peers that had already progressed to a later state (e.g. PeerQueried or PeerWaiting) to regress into PeerHeard and mess with query.

We could also allow PeerUnreachable to become PeerHeard again if we wanted. I think we're probably better off waiting on this though until we have signed peer records implemented and are tracking which addresses have been dialed for a particular peer as part of a query. Could be persuaded either way.

Reply:
@aarshkshah1992 I'm not sure what attack you're thinking of, but it looks like it's an attack vector either way we decide (and I think less if we allow Unreacheable -> Heard)

Enabling PeerUnreachable -> PeerHeard: The attack is an adversary forcing me to waste resources by dialing unreachable peers (relies on a Sybil attack happening to be effective)

Disabling PeerUnreachable -> PeerHeard: The attack is an adversary stopping me from querying other good peers by returning to me the "optimal" peerIDs but with bogus multiaddresses (largely but not entirely alleviated by signed peer records). Then I never query the optimal peers because I've marked them all as obsolete.

While it looks like the attack with Disabling Unreachable -> Heard is worse it also performs worse in the good case where the peer that used to be the closest just goes offline. Now I may have to dial that same peer many times. libp2p dial backoffs could help, but really we should probably be tracking if we learned any new information about a peer before we query it again.